src/workerd/api/streams/internal.c++

```diff
@@ -78,35 +78,125 @@ private:
+    // is well behaved.
+    // If the stream does report a length, once we've read that number of bytes, we'll
+    // fallback to the conservativeAllocation.
+    uint64_t amountToRead = kj::min(limit,
+                                    kj::min(MAX_BUFFER_CHUNK,
+                                            maybeLength.orDefault(DEFAULT_BUFFER_CHUNK)));
+    // amountToRead can be zero if the stream reported a zero-length. While the stream could
+    // be lying about it's length, let's skip reading anything in this case.
     if (amountToRead > 0) {
       for (;;) {
-        // TODO(perf): We can likely further optimize this loop by checking to see
-        // how much of the buffer was filled and using the remaining buffer space if
-        // it is not completely filled by the previous iteration. Doing so makes this
-        // loop a bit more complicated tho, so for now let's keep things simple.
         auto bytes = kj::heapArray<T>(amountToRead);
-        size_t amount = co_await input.tryRead(bytes.begin(), 1, bytes.size());
+        // Note that we're passing amountToRead as the *minBytes* here so the tryRead should
+        // attempt to fill the entire buffer. If it doesn't, the implication is that we read
+        // everything.
+        uint64_t amount = co_await input.tryRead(bytes.begin(), amountToRead, amountToRead);
+        KJ_DASSERT(amount <= amountToRead);
 
-        if (amount == 0) {
+        runningTotal += amount;
+        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
+
+        if (amount < amountToRead) {
+          // The stream has indicated that we're all done by returning a value less than the
+          // full buffer length.
+          // It is possible/likely that at least some amount of data was written to the buffer.
+          // In which case we want to add that subset to the parts list here before we exit
+          // the loop.
+          if (amount > 0) {
+            parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
+          }
           break;
         }
 
-        runningTotal += amount;
-        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
-        parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      };
+        // Because we specify minSize equal to maxSize in the tryRead above, we should only
+        // get here if the buffer was completely filled by the read. If it wasn't completely
+        // filled, that is an indication that the stream is complete which is handled above.
+        KJ_DASSERT(amount == bytes.size());
+        parts.add(kj::mv(bytes));
+
+        // If the stream provided an expected length and our running total is equal to
+        // or greater than that length then we assume we're done.
+        KJ_IF_SOME(length, maybeLength) {
+          if (runningTotal >= length) {
+            // We've read everything we expect to read but some streams need to be read
```

Comment:
I thought we were going to end the loop early here?

Did you observe an actual stream implementation that returned non-null for `tryGetLength()` but had a problem with not being read to EOF? If so we should fix that implementation.

Reply:
Yeah, I had tried but we ended up having some broken tests. I've got a todo to go off and investigate those. I don't think it's actually that those are returning an incorrect `tryGetLength()` as much as they need a final zero-length read to clean up after themselves but I need to verify. And yes, we should absolutely fix those. Once we do, I'll be able to go back and remove this additional extraneous read.

Reply:
@jasnell Ah, I guess I'd argue that if a stream returns a value for `tryGetLength()`, then it should be its responsibility to track when that value hits zero and do the final cleanup then without requiring an additional read. But I can definitely imagine there exist some streams that don't get this right today.